awx/main/models/unified_jobs.py

```diff
@@ -1512,3 +1512,77 @@ def log_lifecycle(self, state, blocked_by=None):
         else:
             msg = f"{self._meta.model_name}-{self.id} {state.replace('_', ' ')}"
         logger_job_lifecycle.debug(msg, extra=extra)
+
+    @property
+    def launched_by(self):
+        ancestor_job = self.ancestor_job
+        launched_by = {}
+        if ancestor_job.launch_type == "manual" or ancestor_job.launch_type == "relaunch":
```

Comment:
There's a lot of duplication in this `if` statement.  You could probably refactor most of this to do e.g.,

```python
if ancestor_job.launch_type in ('manual', 'relaunch'):
    obj = ancestor_job.created_by
elif ancestor_job.launch_type == "scheduled":
    obj = ancestor_job.schedule
elif:
    ...
return {
    'id': obj.id,
    'name': getattr(obj, 'name', None) or obj.username,
    'type': obj.__class__.__name__,
    'url': obj.get_absolute_url()
}

Reply:
This could probably be refined even further with a dict e.g.,

```python
attr = {
    'manual': 'created_by',
    'relaunch': 'created_by',
    'scheduled': 'schedule',
    ...
}
obj = getattr(ancestor_job, attr)
```

Reply:
I concur.

Reply:
More like

    attr = { ... }
    obj = getattr(ancestor_job, attr.get(ancestor_job.launch_type, ''), None)
    if obj is not None:
        return { ... }
    return {}